src/kinetics/ElectronCollisionPlasmaRate.cpp

```diff
@@ -0,0 +1,158 @@
+    pp.getElectronEnergyLevels(levels.data());
+
+    Eigen::ArrayXd dist(pp.nElectronEnergyLevels());
+    pp.getElectronEnergyDistribution(dist.data());
+
+    bool changed = false;
+
+    if (levels.size() != EnergyLevels.size() || (levels - EnergyLevels).any() != 0.0) {
+        EnergyLevels = std::move(levels);
+        changed = true;
+    }
+
+    if (dist.size() != Distribution.size() || (dist - Distribution).any() != 0.0) {
+        Distribution = std::move(dist);
+        changed = true;
+    }
+
+    return changed;
+}
+
+ElectronCollisionPlasmaRate::ElectronCollisionPlasmaRate()
+{
+}
+
+ElectronCollisionPlasmaRate::ElectronCollisionPlasmaRate(const double* energyLevels,
+                                                         const double* crossSection,
+                                                         size_t length)
+{
+    m_energyLevels = Eigen::Map<const Eigen::ArrayXd>(energyLevels, length);
+    m_crossSection = Eigen::Map<const Eigen::ArrayXd>(crossSection, length);
+}
+
+void ElectronCollisionPlasmaRate::setParameters(const AnyMap& node, const UnitStack& rate_units)
+{
+    ReactionRate::setParameters(node, rate_units);
+    if (node.hasKey("energy-levels")) {
+        vector<double> levels = node["energy-levels"].asVector<double>();
+        m_energyLevels = Eigen::Map<const Eigen::ArrayXd>(levels.data(), levels.size());
+    } else {
+        m_energyLevels = NAN;
+        return;
+    }
+    if (node.hasKey("cross-section")) {
+        vector<double> crossSection = node["cross-section"].asVector<double>();
+        m_crossSection = Eigen::Map<const Eigen::ArrayXd>(crossSection.data(), crossSection.size());
+    } else {
+        m_crossSection = NAN;
+        return;
+    }
+    if (m_energyLevels.size() != m_crossSection.size()) {
+        throw CanteraError("ElectronCollisionPlasmaRate::setParameters",
+            "Energy levels and cross section must have the same length.");
+    }
+}
+
+void ElectronCollisionPlasmaRate::getParameters(AnyMap& node) const {
+    node["type"] = type();
+    vector<double> energyLevels(m_energyLevels.size());
+    Eigen::Map<Eigen::ArrayXd>(energyLevels.data(), energyLevels.size()) = m_energyLevels;
+    node["energy-levels"] = energyLevels;
+    vector<double> crossSection(m_crossSection.size());
+    Eigen::Map<Eigen::ArrayXd>(crossSection.data(), crossSection.size()) = m_crossSection;
+    node["cross-section"] = crossSection;
+}
+
+double ElectronCollisionPlasmaRate::evalFromStruct(const ElectronCollisionPlasmaData& shared_data) const
+{
+    Eigen::ArrayXd eps2 = shared_data.EnergyLevels.pow(2.0);
+    Eigen::ArrayXd cross_section(shared_data.EnergyLevels.size());
+    linearInterp(shared_data.EnergyLevels, m_energyLevels, m_crossSection, cross_section);
+    // unit in kmol/m3/s
+    return 0.5 * pow(2.0 * ElectronCharge / ElectronMass, 0.5) * Avogadro *
+           simpson(shared_data.Distribution.cwiseProduct(cross_section), eps2);
```

Comment:
`Eigen`'s `cwiseProduct` (in addition to `pow` above) appears to be the main reason to use `Eigen::ArrayXd`. I believe the implementation would be much clearer if you calculated the data that goes into the integral explicitly. I am pretty sure that `Eigen` is optimized, but I don't believe that the associated complexity is really needed here.

Reply:
My reason of choosing Eigen::ArrayXd is its extensibility to a more complex calculation. In the future, cross-section can be used to calculate the electron energy distribution function. It would be very useful to have the function like cwiseProduct or other array/vector related functions.

Reply:
I think this is legitimate. At the same time, I don't think that it's necessary to have `Eigen::ArrayXd` objects outside of the function itself - you can map memory rather late, and avoid introducing a new `linearInterp` method. There also are some hybrid options, where you allocate memory (including a work vector) as `std::vector` and also keep a copy of a mapped `Eigen::ArrayXd` around.

On a related note: can the `EnergyLevels` vector change within a simulation, or is it static? If these are static (or even semi-static), then I see some opportunities for optimization.

Reply:
The `EnergyLevels` and `cross-sections` will stay constant within a simulation.

Reply:
Thanks - just to make sure: I was referring to `shared_data.EnergyLevels`, not `m_energyLevels`.

Reply:
`shared_data.EnergyLevels` can change during the simulation. `m_energyLevels` will stay the the same.

Reply:
Thank you for the clarification - that’s what I thought but I wanted to make sure.

Regarding my original comment, I still believe it would be preferable if data were stored using standard types (I.e. `vector`) as it’s very simple to map to Eigen whenever it matters. Plus, you avoid the back-and-forth casting in the interpolation method you added, as you can simply use the existing method.